### PR TITLE
setting/keybindings dialog

### PR DESCRIPTION
**Pull request by kasperpeulen**
I haven't tested yet how it looks on mac, will test that tomorrow, when I have my macbook back, also will give the code a second look tomorrow, to clean it up.

@devoncarew I have now included the keys library in dialog.dart. Not sure if you would prefer that I refactor it so that that the keys part is only in playground.dart.

### CODE REVIEW

**Comment by lukechurch** [comment]
@kasperpeulen looks like great progress! Thanks.

For now lets just keep this to the keyboard shortcuts and land that. We can then work out what we feel might be good to surface in options knowing that we have the capability there.

Could you also add a close link to the UI? Probably a small CLOSE hyperlink in the top right?

Ctrl + S doesn't actually do much in DartPad right now, other than stopping the browser showing you a save dialog. Lets take this out of the list.

**Comment by lukechurch** [comment]
@kasperpeulen could you also take a quick look at the build failure please?

File: lib/core/keys.dart [hunk 1]
```diff
@@ -67,6 +66,29 @@ class Keys {
...
+
+class Action {
+  final Function function;
+  final String description;
+
+  Action(this.function, this.description);
+
+  call() => function();
+
+  toString() => description;
```
Comment (devoncarew) [inline lib/core/keys.dart]: `String` toString() ...

File: lib/core/keys.dart [hunk 2]
```diff
@@ -67,6 +66,29 @@ class Keys {
...
+
+  bool operator==(other)
+      => other is Action && description == other.description;
```
Comment (devoncarew) [inline lib/core/keys.dart]: nit: you want to break the line after the `=>`

File: lib/core/keys.dart [hunk 3]
```diff
@@ -90,6 +112,31 @@ String printKeyEvent(KeyboardEvent event) {
...
+  } else {
+    if (keyAsList.any((s) => s == "macctrl")) {
+      return null;
+    }
+    keyAsList = keyAsList.map((String s) => capitalize(s)).toList();
+    return keyAsList.join("+");
+  }
+}
+
+String capitalize(String s) => '${s.substring(0,1).toUpperCase()}${s.substring(1)}';
```
Comment (devoncarew) [inline lib/core/keys.dart]: This is defined twice. Can you move it to somewhere like `src/util.dart`? Also we should cover the cases of empty strings and strings of length 1.

File: lib/dialogs.dart [hunk 4]
```diff
@@ -10,6 +10,8 @@ import 'dart:html';
 import 'elements/elements.dart';
 import 'sharing/gists.dart';
 import 'sharing/mutable_gist.dart';
+import 'package:dart_pad/core/keys.dart';
```
Comment (devoncarew) [inline lib/dialogs.dart]: You can use a relative import here.

File: lib/dialogs.dart [hunk 5]
```diff
@@ -10,6 +10,8 @@ import 'dart:html';
...
+import 'package:dart_pad/dart_pad.dart';
```
Comment (devoncarew) [inline lib/dialogs.dart]: I'd like to just tackle showing keybindings for now; settings can come later, once there's general agreement about how they should be surfaced.

File: lib/dialogs.dart [hunk 6]
```diff
@@ -144,3 +146,59 @@ class SharingDialog extends DDialog {
     });
   }
 }
+
+class SettingsDialog extends DDialog {
```
Comment (devoncarew) [inline lib/dialogs.dart]: This should probably be called something like `KeysDialog`.

File: lib/dialogs.dart [hunk 7]
```diff
@@ -144,3 +146,59 @@ class SharingDialog extends DDialog {
...
+  Map keyMap;
+
+  Map get optionMap => options.values;
+
+  SettingsDialog(this.keyMap) {
+    element.classes.toggle('settings-dialog', true);
+    var text = new ParagraphElement();
+
+    text.text = 'Sharing this pad will create a permanent, publicly visible '
+    'copy on gist.github.com.';
```
Comment (devoncarew) [inline lib/dialogs.dart]: Let's hope not :)

File: lib/dialogs.dart [hunk 8]
```diff
@@ -144,3 +146,59 @@ class SharingDialog extends DDialog {
...
   }
 }
+
+class SettingsDialog extends DDialog {
+
+  Map keyMap;
+
+  Map get optionMap => options.values;
+
+  SettingsDialog(this.keyMap) {
```
Comment (devoncarew) [inline lib/dialogs.dart]: Can you pass a dialog title into the super's ctor?

File: lib/dialogs.dart [hunk 9]
```diff
@@ -144,3 +146,59 @@ class SharingDialog extends DDialog {
...
+    });
+    return dl;
+  }
+
+  String capitalize(String s) => '${s[0].toUpperCase()}${s.substring(1)}';
+
+
+  void show() {
+    super.show();
+    element.children = [];
```
Comment (devoncarew) [inline lib/dialogs.dart]: We should instead add the content to `content` to get a similar look to the other dialogs.

File: web/index.css [hunk 10]
```diff
@@ -385,6 +385,12 @@ div.toggle {
   background: #3f3f3f;
 }
 
+.sharing-dialog {
```
Comment (devoncarew) [inline web/index.css]: I think this needs to be merged with the latest. The sharing dialog isn't positioned any differently than the other ones.

File: lib/core/keys.dart (hunk 1 above)
Comment (kasperpeulen) [inline lib/core/keys.dart]: done

File: lib/core/keys.dart (hunk 2 above)
Comment (kasperpeulen) [inline lib/core/keys.dart]: done

File: lib/core/keys.dart (hunk 3 above)
Comment (kasperpeulen) [inline lib/core/keys.dart]: done

File: lib/dialogs.dart (hunk 4 above)
Comment (kasperpeulen) [inline lib/dialogs.dart]: done

File: lib/dialogs.dart (hunk 5 above)
Comment (kasperpeulen) [inline lib/dialogs.dart]: done

File: web/index.css (hunk 10 above)
Comment (kasperpeulen) [inline web/index.css]: done

File: lib/dialogs.dart (hunk 9 above)
Comment (kasperpeulen) [inline lib/dialogs.dart]: done

File: lib/dialogs.dart (hunk 8 above)
Comment (kasperpeulen) [inline lib/dialogs.dart]: done

File: lib/dialogs.dart (hunk 7 above)
Comment (kasperpeulen) [inline lib/dialogs.dart]: ooopz

File: lib/dialogs.dart (hunk 6 above)
Comment (kasperpeulen) [inline lib/dialogs.dart]: done

**Comment by kasperpeulen** [comment]
Okay, reviewed the comments. Btw, what do you guys think about which options should be surfaced ?

My main motivation for this settings dialog was actually to have a place for an sport++ option (as luke suggested). But as I thought this was also a good place for showing the keybindings, I started with that. As long as we don't have much options/keybindings, I think it makes sense to place keybindings and options both in the same place.

I would also like to expose the experimental options (param and auto completion), as I think it will help if people can test it and report issues. 

Thoughts @devoncarew @lukechurch  ?

This is how the settings now look at a mac btw:

File: lib/playground.dart [hunk 11]
```diff
@@ -278,17 +281,24 @@ class Playground implements GistContainer, GistController {
     _editpanel.children.first.attributes['flex'] = '';
     editor.resize();
 
-    keys.bind(['ctrl-s'], _handleSave);
-    keys.bind(['ctrl-enter'], _handleRun);
+    keys.bind(['ctrl-s'], sharingDialog.show, "Share");
```
Comment (devoncarew) [inline lib/playground.dart]: Let's keep this bound to the no-op 'save' call. User's have muscle memory of what hitting ctrl-s will do; they would have to un-learn it for dartpad.

**Comment by devoncarew** [comment]
Looks great! I think we should change ctrl-s back to 'save', even though save does nothing. Will merge after that change.

Re: exposing advanced options in a UI, I'm a bit on the fence about that. I'd like dartpad to have as simple a UI as possible, even at the expense of making it harder to turn on advanced options. If it's something we really want people to try out, maybe it's important enough to put as a toggle on the main page. Or run an a/b experiment, with some portion of dartpad user's getting the experimental UI?

In any case, we can keep this PR just about key bindings, and noodle more about exposing advanced options.

File: lib/playground.dart (hunk 11 above)
Comment (kasperpeulen) [inline lib/playground.dart]: Okay, makes sense. But shall I delete this from the menu then until we really do some save action when pressing ctrl-s ?

Comment (devoncarew) [inline lib/playground.dart]: I suspect we want this hidden somehow. It's nice to track when the hit ctrl-s, but not critical. Let's comment out the key-binding for now.

**Comment by kasperpeulen** [comment]
fix #295

**Pull request merged**